lrlex/src/lib/parser.rs

```diff
@@ -46,56 +48,104 @@ impl StartState {
 pub(super) struct LexParser<StorageT> {
     src: String,
     pub(super) rules: Vec<Rule<StorageT>>,
-    duplicate_names: HashMap<Span, Vec<Span>>,
     pub(super) start_states: Vec<StartState>,
 }
 
+fn add_duplicate_occurrence(
+    errs: &mut Vec<LexBuildError>,
+    kind: LexErrorKind,
+    orig_span: Span,
+    dup_span: Span,
+) {
+    if !errs.iter_mut().any(|e| {
+        if e.kind == kind && e.spans[0] == orig_span {
+            e.spans.push(dup_span);
+            true
+        } else {
+            false
+        }
+    }) {
+        errs.push(LexBuildError {
+            kind,
+            spans: vec![orig_span, dup_span],
+        });
+    }
+}
+
 impl<StorageT: TryFrom<usize>> LexParser<StorageT> {
     pub(super) fn new(src: String) -> LexBuildResult<LexParser<StorageT>> {
         let mut p = LexParser {
             src,
             rules: Vec::new(),
-            duplicate_names: HashMap::new(),
-            start_states: vec![StartState::new(0, INITIAL_START_STATE_NAME, false)],
+            start_states: vec![StartState::new(
+                0,
+                INITIAL_START_STATE_NAME,
+                false,
+                Span::new(0, 0),
+            )],
         };
         p.parse()?;
         Ok(p)
     }
 
     fn mk_error(&self, kind: LexErrorKind, off: usize) -> LexBuildError {
         let span = Span::new(off, off);
-        LexBuildError { kind, span }
+        LexBuildError {
+            kind,
+            spans: vec![span],
+        }
     }
 
     fn parse(&mut self) -> LexBuildResult<usize> {
-        let mut i = self.parse_declarations(0).map_err(|e| vec![e])?;
-        i = self.parse_rules(i).map_err(|e| vec![e]).and_then(|i| {
-            if let Some((orig_span, spans)) = self.duplicate_names.iter().next() {
-                return Err(vec![LexBuildError {
-                    span: *orig_span,
-                    kind: LexErrorKind::DuplicateName(spans.clone()),
-                }]);
-            }
-            Ok(i)
-        })?;
-        // We don't currently support the subroutines part of a specification. One day we might...
-        match self.lookahead_is("%%", i) {
-            Some(j) => {
-                let k = self.parse_ws(j).map_err(|e| vec![e])?;
-                if k == self.src.len() {
-                    Ok(i)
-                } else {
-                    Err(vec![self.mk_error(LexErrorKind::RoutinesNotSupported, i)])
+        let mut errs = Vec::new();
+        let mut result = self.parse_declarations(0, &mut errs);
+        result = self.parse_rules(
+            match result {
```

Comment:
As we're NOT parsing rules when the `parse_declarations` returns an error, would it not be more legible to pull out this match - possibly into something like
```
if let Err(e) = result {
   errs.push(e);
   return Err(errs);
} else {
  result = self.parse_rules(result.unwrap(), // unwrap safe due to the above
               &mut errs,
  );
}

Reply:
That is an option, I mostly went with the match to avoid `unwrap()`, but another option instead of embedding the match in `parse_rules` we could

```rust
let mut i = match self.parse_declarations(0) {
        Ok(i) => i,
        Err(e) => {
           errs.push(e); 
           return Err(errs);
        }
};
```

It ended up this primarily as a consequence of the way `YaccGrammarError` was at the start of a series,
with things happening between `result = parse_foo(..);` and the subsequent `match`, but that has been refactored out of `YaccGrammarError`, and isn't the case here at all.  I just hadn't reconsidered it since the start of that other series.

Reply:
Definitely better than the embedded match

Reply:
I implemented that in https://github.com/softdevteam/grmtools/pull/330/commits/5f6a164f261ef08fe42b74afdb6ded8ab6f9986c along with the other issues I noticed in the `parse` function.